File: objc/Quick Start/MapViewController.m
//
//  ViewController.m
//  Quick Start
//
//  Created by Alex Pavlov on 6/18/15.
//  Copyright (c) 2015 Skyhook Wireless. All rights reserved.
//

#import "MapViewController.h"

@interface MapViewController ()

@property (weak, nonatomic) IBOutlet MKMapView *mapView;
@property (nonatomic, strong) CLLocationManager *locationManager;
@property (weak, nonatomic) SHXAccelerator *accelerator;

@end

@implementation MapViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    self.mapView.delegate = self;
    self.locationManager = [CLLocationManager new];
    self.locationManager.delegate = self;
    self.accelerator = ((AppDelegate*)[UIApplication sharedApplication].delegate).accelerator;
    
    if (self.accelerator != nil)
    {
        self.accelerator.delegate = self;
    }
}

-(void)viewDidAppear:(BOOL)animated
{
    [super viewDidAppear:animated];
    [self.accelerator startMonitoringForAllCampaigns];
    
    // Actually, we are not going to show user location on the map. Our goal is to zoom map view
    // to current location. As soon as mapView:didUpdateUserLocation receives first coordinate, we
    // will set showsUserLocation to NO
    self.mapView.showsUserLocation =
        ([CLLocationManager authorizationStatus] == kCLAuthorizationStatusAuthorizedAlways);
}

#pragma mark - CLLocationManagerDelegate

-(void)locationManager:(CLLocationManager *)manager didChangeAuthorizationStatus:(CLAuthorizationStatus)status
{
        self.mapView.showsUserLocation = (status == kCLAuthorizationStatusAuthorizedAlways);
}

#pragma mark - SHXAcceleratorDelegate

-(void)accelerator:(SHXAccelerator *)accelerator didFailWithError:(NSError *)error
{
    NSLog(@"accelerator didFailWithError %@", error);
    
    if (error.code == SHXErrorRegionMonitoringUnavailable)
    {
        UIAlertController* alert = [UIAlertController alertControllerWithTitle:@"Device not supported"
                                                                       message:@"Skyhook SDK requires geofencing, which is not available on your device."
                                                                preferredStyle:UIAlertControllerStyleAlert];
        
        // Keep in mind that we assume a single view app here.
        [self presentViewController:alert animated:YES completion:nil];
    }
}

-(void)accelerator:(SHXAccelerator *)accelerator didEnterVenue:(SHXCampaignVenue *)venue
{
    // SHXAcceleratorDelegate methods are always executed on main thread. Depending on
    // how much work your code is going to do you might want to run it off main thread.
    NSLog(@"accelerator venue entry: %@ %@", venue, venue.venueIdent);
    
    [accelerator fetchInfoForVenues:@[venue.venueIdent] completion:^(NSArray *venueInfoList, NSError *error)
     {
         if (venueInfoList)
         {
             SHXVenueInfo *venueInfo = venueInfoList[0];
             UILocalNotification *notification = [UILocalNotification new];
             notification.fireDate = nil;
             notification.alertBody = [NSString stringWithFormat:@"Approaching %@", venueInfo.placemark.name];
             notification.soundName = UILocalNotificationDefaultSoundName;
             [[UIApplication sharedApplication] scheduleLocalNotification:notification];
             
             MKPointAnnotation *annotation = [MKPointAnnotation new];
             annotation.coordinate = venueInfo.placemark.coordinate;
             annotation.title = venueInfo.placemark.name;
             [self.mapView addAnnotation:annotation];
         }
         else
         {
             NSLog(@"Error: %@", error);
         }
     }];
}


#pragma mark - MKMapViewDelegate

-(void)mapView:(MKMapView *)mapView didUpdateUserLocation:(MKUserLocation *)userLocation
{
    mapView.showsUserLocation = NO;
    MKCoordinateRegion coordinateRegion = MKCoordinateRegionMakeWithDistance(userLocation.location.coordinate, 1000, 1000);
    [self.mapView setRegion:coordinateRegion animated:YES];
}

- (MKAnnotationView *)mapView:(MKMapView *)mapView viewForAnnotation:(id <MKAnnotation>)annotation
{
    if ([annotation isKindOfClass:[MKPointAnnotation class]])
    {
        static NSString *kReuseId = @"VenuePin";
        MKPinAnnotationView *pinView = (MKPinAnnotationView *) [self.mapView dequeueReusableAnnotationViewWithIdentifier:kReuseId];
        
        if (pinView == nil)
        {
            pinView = [[MKPinAnnotationView alloc] initWithAnnotation:annotation reuseIdentifier:kReuseId];
            pinView.canShowCallout = YES;
        }
        return pinView;
    }
    return nil;
}


@end

File: swift/QuickStartSwift/MapViewController.swift
//
//  MapViewController.swift
//  QuickStartSwift
//
//  Created by Alex Pavlov on 5/29/18.
//  Copyright © 2018 Skyhook. All rights reserved.
//

import UIKit
import UserNotifications
import MapKit
import CoreLocation
import SkyhookContext

class MapViewController: UIViewController, MKMapViewDelegate, CLLocationManagerDelegate, SHXAcceleratorDelegate {

    static let kReuseId = "VenuePin"
    
    @IBOutlet weak var mapView: MKMapView!
    let locationManager = CLLocationManager()
    
    var accelerator: SHXAccelerator? {
        get {
            return (UIApplication.shared.delegate as? AppDelegate)?.accelerator
        }
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        mapView.delegate = self
        locationManager.delegate = self
        
        if let accelerator = accelerator {
            accelerator.delegate = self
        }
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        accelerator?.startMonitoringForAllCampaigns()
        
        // Actually, we are not going to show user location on the map. Our goal is to zoom map view
        // to current location. As soon as mapView:didUpdateUserLocation receives first coordinate, we
        // will set showsUserLocation to false
        mapView.showsUserLocation = (CLLocationManager.authorizationStatus() == .authorizedAlways)
    }
    

    // MARK: - CLLocationManagerDelegate
    
    func locationManager(_ manager: CLLocationManager, didChangeAuthorization status: CLAuthorizationStatus) {
        mapView.showsUserLocation = (status == .authorizedAlways)
    }
    
    // MARK: - SHXAcceleratorDelegate
    
    func accelerator(_ accelerator: SHXAccelerator!, didFailWithError error: Error!) {
        
        
        if error._code == SHXError.regionMonitoringUnavailable.rawValue {
            let alert = UIAlertController(title: "Device not supported",
                                          message: "Skyhook SDK requires geofencing, which is not available on your device.",
                                          preferredStyle: .alert)
            present(alert, animated: true, completion: nil)
        }
    }
    
    func accelerator(_ accelerator: SHXAccelerator!, didEnter venue: SHXCampaignVenue!) {
        
        // SHXAcceleratorDelegate methods are always executed on main thread. Depending on
        // how much work your code is going to do you might want to run it off main thread.
        NSLog("accelerator venue entry: \(venue!) \(venue!.venueIdent!)")
        
        accelerator.fetchInfo(forVenues: [venue!.venueIdent!]) { [weak self] (venues, error) in
            
            if let error = error {
                NSLog("Error: \(error)")
                return
            }
            
            // Post notification
            let venueInfo: SHXVenueInfo = venues![0] as! SHXVenueInfo
            let content = UNMutableNotificationContent()
            content.body = "Approaching \(venueInfo.placemark.name ?? "")"
            content.sound = UNNotificationSound.default()
            let request = UNNotificationRequest(identifier: "DemoNotification-\(venue!.venueIdent!)", content: content, trigger: nil)
            UNUserNotificationCenter.current().add(request, withCompletionHandler: nil)
            
            // Add map annotation
            let annotation = MKPointAnnotation()
            annotation.coordinate = venueInfo.placemark.coordinate
            annotation.title = venueInfo.placemark.name
            self?.mapView.addAnnotation(annotation)
        }
    }
    

    // MARK: - MKMapViewDelegate
    
    func mapView(_ mapView: MKMapView, didUpdate userLocation: MKUserLocation) {
        mapView.showsUserLocation = false
        let coordinateRegion = MKCoordinateRegionMakeWithDistance(userLocation.coordinate, 1000, 1000)
        mapView.setRegion(coordinateRegion, animated: true)
    }
    
    func mapView(_ mapView: MKMapView, viewFor annotation: MKAnnotation) -> MKAnnotationView? {
        
        if let pointAnnotation: MKPointAnnotation = annotation as? MKPointAnnotation {
            return mapView.dequeueReusableAnnotationView(withIdentifier: MapViewController.kReuseId) ?? buildPinView(pointAnnotation)
        }
        return nil
    }
    
    func buildPinView(_ annotation: MKPointAnnotation) -> MKPinAnnotationView
    {
        let pinView = MKPinAnnotationView(annotation: annotation, reuseIdentifier: MapViewController.kReuseId)
        pinView.canShowCallout = true
        return pinView
    }
    
}


